src/python/dxpy/nextflow/nextflow_builder.py

```diff
@@ -0,0 +1,158 @@
+
+    Runs the Nextflow Pipeline Importer app, which creates NF applet from given repository.
+    """
+    # FIXME: is this already present somewhere?
+    def create_dxlink(dx_object):
+        try:
+            if dxpy.is_dxlink(dx_object):
+                return dx_object
+            if ":" in dx_object:
+                object_project, object_id = dx_object.split(":", 1)
+            else:
+                object_id = dx_object
+                object_project = None
+            if not dxpy.utils.resolver.is_hashid(object_id):
+                object_project, _, object_id = dxpy.utils.resolver.resolve_existing_path(object_id, expected="entity", expected_classes=["file"], describe=False)
+                object_id = object_id["id"]
+            return dxpy.dxlink(object_id=object_id, project_id=object_project)
+        except dxpy.utils.resolver.ResolutionError:
+            print("GitHub credentials ('{}') file could not be found!".format(dx_object), file=sys.stderr)
+            exit(2)
+
+
+    build_project_id = dxpy.WORKSPACE_ID
+    if build_project_id is None:
+        parser.error(
+            "Can't create an applet without specifying a destination project; please use the -d/--destination flag to explicitly specify a project")
+
+    input_hash = {
+        "repository_url": repository,
+    }
+    if tag:
+        input_hash["repository_tag"] = tag
+    if profile:
+        input_hash["config_profile"] = profile
+    if github_creds:
+        input_hash["github_credentials"] = create_dxlink(github_creds)
+
+    api_options = {
+        "name": "Nextflow build of %s" % (repository),
+        "input": input_hash,
+        "project": build_project_id,
+    }
+    try:
+        app_run_result = dxpy.api.app_run('app-nextflow_pipeline_importer', input_params=api_options)
+    except dxpy.exceptions.ResourceNotFound:
+        print("GitHub credentials file ('{}') could not be found!".format(github_creds), file=sys.stderr)
+        raise
+    job_id = app_run_result["id"]
+    if not brief:
+        print("Started builder job %s" % (job_id,))
+    dxpy.DXJob(job_id).wait_on_done(interval=1)
+    applet_id, _ = dxpy.get_dxlink_ids(dxpy.api.job_describe(job_id)['output']['output_applet'])
+    if not brief:
+        print("Created Nextflow pipeline %s" % (applet_id))
+    else:
+        print(applet_id)
+    return applet_id
+
+def prepare_nextflow(resources_dir, profile):
+    """
+    :param resources_dir: Directory with all resources needed for Nextflow Pipeline. Usually directory with user's NF files.
+    :type resources_dir: str or Path
+    :param profile: Custom NF profile, for more information visit https://www.nextflow.io/docs/latest/config.html#config-profiles
+    :type profile: string
+
+    Creates files for creating applet, such as dxapp.json and source file. These files are created in temp directory.
+    """
+    assert os.path.exists(resources_dir)
+    inputs = []
+    os.makedirs(".dx.nextflow", exist_ok=True)
```

Comment:
Seems that we are not using this `./.dx.workflow`, right?

correct me if I am wrong but perhaps we can do this:
if the user call `dx build some/path/to/nf-hello` where the `main.nf` and its own files are located:
1. mkdir `./.dx.nextflow` in cwd
2. `get_nextflow_dxapp` to make `dxapp.json` , and store it in `./.dx.nextflow` (instead of in the template dir, so the templates wont get changed)
3. `get_nextflow_src`vto make `nextflow.sh`, and store it in `./.dx.nextflow/src` as @commandlinegirl suggested
4. copy the `resources/usr/local/bin/*` from the template dir to `./dx.nextflow/resources/usr/local/bin`
5. copy the user source `some/path/to/nf-hello/*` to `./.dx.nextflow/resources/home/dnanexus/nxf`
6. return `./.dx.nextflow` to the build_applet, so we use `./.dx.nextflow` as the actual applet src folder as we usually do when building other applet.

and what we have in `./.dx.nextflow` would be:
dxapp.json
src/nextflow.sh
resources/usr/local/bin from temp dir
resources/home/dnanexus/nxf from the user given source folder and has all nxf needed files

and when the applet get built, the bin folder from template dir goes to the `/usr/local/bin` and the main.nf etc. go to `/home/dnanexus/nxf`
@mhrvol

Reply:
Good call. You are right copying could be really labor intensive. Now I get your point. Thanks and nice work :D